### PR TITLE
msg/async: increase worker reference with local listen table enabled backend

### PR DESCRIPTION
**Pull request by yuyuyu101**
Signed-off-by: Haomai Wang <haomai@xsky.com>

### CODE REVIEW

File: src/msg/async/AsyncMessenger.cc [hunk 1]
```diff
@@ -171,13 +171,16 @@ void Processor::accept()
     Worker *w = worker;
     if (!msgr->get_stack()->support_local_listen_table())
       w = msgr->get_stack()->get_worker();
+    else
+      ++w->references;
     int r = listen_socket.accept(&cli_socket, opts, &addr, w);
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.cc]: is `w` used/held anywhere by the `listen_socket.accept()`?

Comment (yuyuyu101) [inline src/msg/async/AsyncMessenger.cc]: yes, worker should be existed always

Comment (yuyuyu101) [inline src/msg/async/AsyncMessenger.cc]: the reference doesn't mean the lifecycle of Worker, it's just a hit about connection aggresive

File: src/msg/async/AsyncMessenger.cc [hunk 2]
```diff
@@ -171,13 +171,16 @@ void Processor::accept()
...
     if (r == 0) {
       ldout(msgr->cct, 10) << __func__ << " accepted incoming on sd " << cli_socket.fd() << dendl;
 
       msgr->add_accept(w, std::move(cli_socket), addr);
       continue;
     } else {
+      --w->references;
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.cc]: can we use `release_worker()` instead?

File: src/msg/async/AsyncMessenger.cc [hunk 3]
```diff
@@ -171,13 +171,16 @@ void Processor::accept()
     Worker *w = worker;
     if (!msgr->get_stack()->support_local_listen_table())
       w = msgr->get_stack()->get_worker();
+    else
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.cc]: can we put 
```c++
if (msgr->get_stack()->support_local_listen_table()) {
  w = worker;
  w->reference;
} else {
  w = msgr->get_stack()->get_worker();
}
```

1. use positive logic to help with the readablity
1. more explicit on what the `w` is assigned
1. post-increment is faster for `atmic<>` types

**Comment by tchaikov** [comment]
retest this please.

File: src/msg/async/AsyncMessenger.cc [hunk 4]
```diff
@@ -165,19 +165,24 @@ void Processor::accept()
...
   opts.priority = msgr->get_socket_priority();
+  Worker *w;
   while (true) {
     entity_addr_t addr;
     ConnectedSocket cli_socket;
-    Worker *w = worker;
-    if (!msgr->get_stack()->support_local_listen_table())
+    if (msgr->get_stack()->support_local_listen_table()) {
+      w = worker;
+      w->reference++;
```
Comment (tchaikov) [inline src/msg/async/AsyncMessenger.cc]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/msg/async/AsyncMessenger.cc: In member function ‘void Processor::accept()’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/msg/async/AsyncMessenger.cc:174:10: error: ‘class Worker’ has no member named ‘reference’
       w->reference++;
```

`s/reference/references/`

Comment (yuyuyu101) [inline src/msg/async/AsyncMessenger.cc]: done

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-07-05_03:23:34-rados-wip-kefu-testing-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-07-04_08:30:19-rados-wip-kefu-testing-distro-basic-mira/

**Pull request merged**